annotated_test.go

```diff
@@ -65,6 +67,196 @@ func TestAnnotated(t *testing.T) {
 	})
 }
 
+type asStringer struct {
+	name string
+}
+
+func (as *asStringer) String() string {
+	return as.name
+}
+
+func TestAnnotatedAs(t *testing.T) {
+	t.Parallel()
+	type in struct {
+		fx.In
+
+		S fmt.Stringer `name:"goodStringer"`
+	}
+	type myStringer interface {
+		String() string
+	}
+
+	newAsStringer := func() *asStringer {
+		return &asStringer{
+			name: "a good stringer",
+		}
+	}
+
+	tests := []struct {
+		desc    string
+		provide fx.Option
+		invoke  interface{}
+	}{
+		{
+			desc: "provide a good stringer",
+			provide: fx.Provide(
+				fx.Annotate(newAsStringer, fx.As(new(fmt.Stringer))),
+			),
+			invoke: func(s fmt.Stringer) {
+				assert.Equal(t, s.String(), "a good stringer")
+			},
+		},
+		{
+			desc: "provide with multiple types As",
+			provide: fx.Provide(fx.Annotate(func() (*asStringer, *bytes.Buffer) {
+				buf := make([]byte, 1)
+				b := bytes.NewBuffer(buf)
+				return &asStringer{name: "stringer"}, b
+			}, fx.As(new(fmt.Stringer), new(io.Writer)))),
+			invoke: func(s fmt.Stringer, w io.Writer) {
+				w.Write([]byte(s.String()))
+			},
+		},
+		{
+			desc: "provide as with result annotation",
+			provide: fx.Provide(
+				fx.Annotate(func() *asStringer {
+					return &asStringer{name: "stringer"}
+				},
+					fx.ResultTags(`name:"goodStringer"`),
+					fx.As(new(fmt.Stringer))),
+			),
+			invoke: func(i in) {
+				assert.Equal(t, "stringer", i.S.String())
+			},
+		},
+		{
+			// same as the test above, except now we annotate
+			// it in a different order.
+			desc: "provide as with result annotation, in different order",
+			provide: fx.Provide(
+				fx.Annotate(func() *asStringer {
+					return &asStringer{name: "stringer"}
+				},
+					fx.As(new(fmt.Stringer)),
+					fx.ResultTags(`name:"goodStringer"`)),
+			),
+			invoke: func(i in) {
+				assert.Equal(t, "stringer", i.S.String())
+			},
+		},
+		{
+			desc: "provide multiple constructors annotated As",
+			provide: fx.Provide(
+				fx.Annotate(func() *asStringer {
+					return &asStringer{name: "stringer"}
+				}, fx.As(new(fmt.Stringer))),
+				fx.Annotate(func() *bytes.Buffer {
+					buf := make([]byte, 1)
+					return bytes.NewBuffer(buf)
+				}, fx.As(new(io.Writer))),
+			),
+			invoke: func(s fmt.Stringer, w io.Writer) {
+				assert.Equal(t, "stringer", s.String())
+				_, err := w.Write([]byte{1})
+				require.NoError(t, err)
+			},
+		},
+		{
+			desc: "provide the same provider as multiple types",
```

Comment:
Would it be possible to design this so it's possible to provide it as multiple interfaces in a si gle provide call, or does that go against the overall design?

We tend to write smaller interfaces, but sometimes larger implementations, which could implement 3-5 interfaces. We would probably end up making a wrapper for it, instead of manually providing the same constructor 5 times.

Reply:
@Multiply thanks for the feedback, and to answer your question, yes. 

A possible workaround for doing what you described is something like the following where `Buffer` implements `io.Writer` and `io.Reader`:
```go
func newBuffer() (*Buffer, *Buffer) {
  b := newBuffer()
  return b, b
}

fx.Annotate(newBuffer, fx.As(new(io.Writer), new(io.Reader)))
```

But that wrapper func is annoying and I agree it'd be nice to not have to do this; in fact, @abhinav and I discussed this earlier today and while it's out of scope for this PR, we're currently thinking it'd look something like:
```go
fx.Provide(
  fx.Annotate(newBuffer, fx.As(new(io.Writer)), fx.As(new(io.Reader)))
)
```

Currently if you try to annotate a constructor with multiple `As`, it'll error out. But we can change that to be allowed once we have the implementation detail sorted out in a near-term future release.

Reply:
That's also the design I had in mind when first looking at it, but didn't seem it was supported in this PR, as you just mentioned. 👍 

Feel free to tag me in the follow up PR, I'll be happy to test the functionality in our codebase, as this is one of the main feature requests from our team! 🚀

Reply:
will do :)

Reply:
@Multiply just FYI, https://github.com/uber-go/fx/pull/817 implements the change we talked about here. Thought I should let you know :)

Reply:
Awesome! I'll try it out right away 🚀